maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,482 @@
+        self._vm_item_picker = self._vm_reader.items_tick_picker(self._start_tick, self._max_tick, time_unit="s")
+
+        self._cpu_reader = CpuReader(data_path=self._config.CPU_READINGS, start_tick=self._start_tick)
+
+        self._tick: int = 0
+
+    @property
+    def configs(self) -> dict:
+        """dict: Current configuration."""
+        return self._config
+
+    @property
+    def frame(self) -> FrameBase:
+        """FrameBase: Current frame."""
+        return self._frame
+
+    @property
+    def snapshots(self) -> SnapshotList:
+        """SnapshotList: Current snapshot list."""
+        return self._snapshots
+
+    def _load_configs(self):
+        """Load configurations."""
+        # Update self._config_path with current file path.
+        self.update_config_root_path(__file__)
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._config = DottableDict(safe_load(fp))
+
+        self._delay_duration: int = self._config.DELAY_DURATION
+        self._buffer_time_budget: int = self._config.BUFFER_TIME_BUDGET
+        self._pm_amount: int = self._config.PM["P1"]["AMOUNT"]
+
+    def _init_data(self):
+        vm_table_data_path = self._config.VM_TABLE
+        if vm_table_data_path.startswith("~"):
+            vm_table_data_path = os.path.expanduser(vm_table_data_path)
+
+        cpu_readings_data_path = self._config.CPU_READINGS
+        if cpu_readings_data_path.startswith("~"):
+            cpu_readings_data_path = os.path.expanduser(cpu_readings_data_path)
+
+        if (not os.path.exists(vm_table_data_path)) or (not os.path.exists(cpu_readings_data_path)):
+            self._download_processed_data()
+
+    def _init_pms(self):
+        """Initialize the physical machines based on the config setting. The PM id starts from 0."""
+        self._pm_cpu_cores_capacity: int = self._config.PM["P1"]["CPU"]
+        self._pm_memory_capacity: int = self._config.PM["P1"]["MEMORY"]
+
+        # TODO: Improve the scalability. Like the use of multiple PM sets.
+        self._machines = self._frame.pms
+        for pm_id in range(self._pm_amount):
+            pm = self._machines[pm_id]
+            pm.set_init_state(
+                id=pm_id,
+                cpu_cores_capacity=self._pm_cpu_cores_capacity,
+                memory_capacity=self._pm_memory_capacity
+            )
+
+    def reset(self):
+        """Reset internal states for episode."""
+        self._total_energy_consumption: float = 0.0
+        self._successful_placement: int = 0
+        self._successful_completion: int = 0
+        self._failed_placement: int = 0
+        self._total_latency: Latency = Latency()
+        self._total_oversubscriptions: int = 0
+
+        for pm in self._machines:
+            pm.reset()
+
+        self._live_vms: Dict[int, VirtualMachine] = {}
+        self._pending_vm_request_payload: Dict[int, VmRequestPayload] = {}
+
+        self._frame.reset()
```

Comment:
reset frame and snapshot before reset machines, or frame.reset will clear all the attribute value of machines